server/src/main/java/org/candlepin/async/JobMessageDispatcher.java

```diff
@@ -14,13 +14,159 @@
+     *
+     * @param cpmSessionFactory
+     *  A CPMSessionFactory instance for creating messaging sessions
+     *
+     * @param objMapper
+     *  An ObjectMapper instance for serializing job messages prior to sending them to the
+     *  message bus
+     */
+    @Inject
+    public JobMessageDispatcher(Configuration config, CPMSessionFactory cpmSessionFactory,
+        ObjectMapper objMapper) {
+
+        this.config = Objects.requireNonNull(config);
+        this.cpmSessionFactory = Objects.requireNonNull(cpmSessionFactory);
+        this.objMapper = Objects.requireNonNull(objMapper);
+
+        this.initialized = false;
+    }
+
+    /**
+     * Initializes this message dispatcher, specfiying the job manager that will be dispatching
+     * messages
+     *
+     * @param manager
+     *  the JobManager that will be dispatching messages
+     *
+     * @throws IllegalStateException
+     *  if this message receiver has already been initialized
+     *
+     * @throws IllegalArgumentException
+     *  if the provided job manager is null
+     */
+    public synchronized void initialize(JobManager manager) {
+        if (this.initialized) {
+            throw new IllegalStateException("Message receiver already initialized");
+        }
+
+        if (manager == null) {
+            throw new IllegalArgumentException("manager is null");
+        }
+
+        // Nothing fancy to do here
+        this.initialized = true;
+    }
+
+    /**
+     * Shuts down this job message receiver, closing any sessions it may have opened
+     */
+    public synchronized void shutdown() throws JobException {
+        try {
+            if (this.cpmSession != null) {
+                this.cpmSession.close();
+            }
+        }
+        catch (CPMException e) {
+            throw new JobException(e);
+        }
+    }
+
+    /**
+     * Checks if this message receiver has been initialized.
+     *
+     * @return
+     *  true if the message receiver has been initialized; false otherwise.
+     */
+    public synchronized boolean isInitialized() {
+        return this.initialized;
+    }
+
+    /**
+     * Fetches the current CPM session, creating a new one if necessary.
+     *
+     * @return
+     *  a CPMSession instance
+     */
+    private CPMSession getSession() throws CPMException {
+        if (this.cpmSession == null || this.cpmSession.isClosed()) {
+            CPMSessionConfig config = this.cpmSessionFactory.createSessionConfig()
+                .setTransactional(true);
+
+            // TODO: Add any other job-system-specific session configuration here
+
+            this.cpmSession = this.cpmSessionFactory.createSession(config);
+            this.cpmSession.start();
+        }
+
+        return this.cpmSession;
+    }
+
+    /**
+     * Fetches the current CPM producer, creating a new one if necessary.
+     *
+     * @return
+     *  a CPMProducer instance
+     */
+    private CPMProducer getProducer() throws CPMException {
```

Comment:
Lots of other methods in this class are synchronized. Does this method need to be too?

Reply:
Yes